app/ts/store/tokens.ts

```diff
@@ -0,0 +1,174 @@
+import { signal, useComputed, useSignal, useSignalEffect } from '@preact/signals'
+import { useAccount } from './account.js'
+import { useAsyncState } from '../library/preact-utilities.js'
+import { useProviders } from './provider.js'
+import { useNetwork } from './network.js'
+import { errors, ethers, logger, utils } from 'ethers'
+import { ERC20ABI } from '../library/ERC20ABI.js'
+
+const CACHEID_PREFIX = '_ut'
+
+const tokens = signal<TokenMeta[]>([])
+
+export function useAccountTokens() {
+	const { address } = useAccount()
+
+	const cacheKey = useComputed(() => {
+		if (address.value.state !== 'resolved') return `${CACHEID_PREFIX}:default`
+		return `${CACHEID_PREFIX}:${address.value.value}`
+	})
+
+	const addToken = (token: TokenMeta) => {
+		tokens.value = [...tokens.value, token]
+	}
+
+	const removeToken = (address: TokenMeta['address']) => {
+		tokens.value = [...tokens.value.filter(token => token.address !== address)]
+	}
+
+	const listenForCacheKeyChange = () => {
+		const tokensCache = useTokensCache(cacheKey.value)
+		if (tokensCache.error !== undefined) throw new Error('Cache data could not be read.')
+		tokens.value = tokensCache.data
+	}
+
+	const listenForTokensChange = () => {
+		if (cacheKey.value === `${CACHEID_PREFIX}:default`) return
+		const newCache = JSON.stringify(tokens.value)
+		localStorage.setItem(cacheKey.value, newCache)
+	}
+
+	useSignalEffect(listenForCacheKeyChange)
+	useSignalEffect(listenForTokensChange)
+
+	return {
+		tokens,
+		addToken,
+		removeToken,
+	}
+}
+
+export function useTokenQuery() {
+	const { value: query, waitFor, reset } = useAsyncState<TokenMeta>()
+	const providers = useProviders()
+	const { network } = useNetwork()
+	const tokenAddress = useSignal('')
+
+	const validateChangedAddress = () => {
+		reset()
+
+		// check address validity
+		if (!utils.isAddress(tokenAddress.value)) return
+
+		waitFor(async () => {
+			if (network.value.state !== 'resolved') logger.throwError('Disconnected', errors.NETWORK_ERROR)
+			const chainId = network.value.value.chainId
+
+			try {
+				const provider = providers.getbrowserProvider()
+				const contract = new ethers.Contract(tokenAddress.value, ERC20ABI, provider)
+				const name = await contract.name()
+				const symbol = await contract.symbol()
+				const decimals = await contract.decimals()
+				return { chainId, name, symbol, decimals, address: tokenAddress.value } as const
+			} catch (unknownError) {
+				logger.throwError('Contract call failed.', errors.CALL_EXCEPTION)
+			}
+		})
+	}
+
+	useSignalEffect(validateChangedAddress)
+
+	return { query, tokenAddress }
+}
+
+// Move to constants later
+export type TokenMeta = {
+	chainId: number
+	name: string
+	address: string
+	symbol: string
+	decimals: number
+}
+
+export const DEFAULT_TOKENS: TokenMeta[] = [
+	{
+		chainId: 1,
```

Comment:
image is missing from here?

Reply:
image icon will be loaded as publicPath + lowecased contract address + file extension else will show default ethereum icon